### PR TITLE
feat: link connect-style server handlers to shared RPC resource nodes

### PR DESCRIPTION
**Pull request by vitali87**
## Summary

Issue #912 slice 2: the server side of generated-RPC linking. A Go wiring call `path, handler := <pkg>connect.New<Stem>Handler(impl)` proves that the impl type serves the generated contract, so each contract method the impl defines now gets an `EXPOSES` edge to the same unscoped `resource::RPC::<Stem>.<Method>` node that client sinks (slice 1, #927/#928/#929) already target. Callers and servers of an interior RPC meet on one resource node, across projects, with no URL matching involved.

## How it works

- The call pass detects `New<Stem>Handler(arg)` calls whose qualifier resolves through the import map to a generated package (name ending in `connect`), mirroring the client-side constructor recognition. A qualifier shadowed by a local is rejected, mirroring the client guard.
- The impl argument resolves to its type by shape: a literal-typed local or inline `&Impl{}`, or a constructor binding (`uSrv := server.New(...)`, `grpcSrv, err := grpc.New(...)`, inline `server.New()`) through the imported package's recorded return type. An unaliased import whose source qualifier is the target's `package` clause (which can differ from the directory name) falls back to clause matching; an ambiguous clause stays silent rather than guessing.
- Contract methods are read syntactically from the vendored codegen's `<Stem>Handler` interface (interface method specs are not graph nodes). Only methods the impl actually defines emit edges, so no dangling references are possible, and everything is gated behind the `io` capture group's `EXPOSES` relationship.
- Package membership uses the (directory, package clause) model from #931 throughout.

## Testing

Nine pipeline tests cover the two verified production wiring shapes (aliased import + cross-package constructor; direct import + multi-return constructor), inline composite-literal and inline constructor arguments, clause-named packages, shadowed qualifiers, ambiguous clause imports, non-connect `New<Stem>Handler` lookalikes, and contract methods the impl does not define. All were written RED first. Full suite: 6069 passed, 10 skipped.

## Summary by CodeRabbit

* **New Features**
  * Added Go connect-go server wiring detection to identify implemented handler methods.
  * Automatically creates corresponding `resource::RPC` endpoint nodes and links them via `EXPOSES` relationships.
  * Supports common wiring patterns including constructor calls, composite literals, aliases, and multi-return bindings.
* **Bug Fixes**
  * Improved Go analysis for promoted/embedded method resolution, correctly excluding unimplemented stubs and handling shadowing and ambiguous imports.
* **Tests**
  * Expanded Go RPC exposure coverage for edge cases like decoy bindings, inline wiring, and embedding graph traversal.

### CODE REVIEW

File: codebase_rag/parsers/rpc_exposure.py [hunk 1]
```diff
@@ -0,0 +1,315 @@
...
+        identity = f"{stem}{cs.SEPARATOR_DOT}{method}"
+        resource_qn = RESOURCE_QN_FORMAT.format(
+            kind=ResourceKind.RPC.value, identity=identity
+        )
+        self._ingestor.ensure_node_batch(
+            cs.NodeLabel.RESOURCE,
+            {
+                cs.KEY_QUALIFIED_NAME: resource_qn,
+                cs.KEY_NAME: identity,
+                KEY_KIND: ResourceKind.RPC.value,
```
[the remark this answers is not in the document]
Comment (vitali87) [reply codebase_rag/parsers/rpc_exposure.py:591]: Fixed in bbf74093. `_method_source_qn` now walks the impl type's embedded fields (read from the struct AST, since Go embedding is not inheritance and the graph has no parent edge), emitting the EXPOSES edge from the promoted method's own node, cycle-safe and depth-bounded. Promoted stubs from a generated `*connect` package (`Unimplemented<Stem>Handler`) are excluded by package clause, since they are not served RPCs. RED tests `test_promoted_method_from_embedded_base_exposes` (your T-Rex repro shape: edge now anchors on `Base.GetUser`) and `test_unimplemented_embed_does_not_expose_stubs` cover both directions.

File: codebase_rag/parsers/rpc_exposure.py [hunk 2]
```diff
@@ -0,0 +1,315 @@
...
+            caller_node, module_qn
+        )
+        call_bindings = self._go_engine.collect_call_var_bindings(caller_node)
+        # A local shadowing the imported package name makes the call a method
+        # on the local value, not codegen wiring (mirrors the client guard).
+        shadowed = set(var_types) | {name for name, _segments in call_bindings}
+        for qualifier, stem, arg in wirings:
+            if qualifier in shadowed:
+                continue
+            impl_qn = self._resolve_impl_qn(module_qn, arg, var_types, call_bindings)
```
[the remark this answers is not in the document]
Comment (vitali87) [reply codebase_rag/parsers/rpc_exposure.py]: Fixed in bbf74093. Resolution is now position-aware: `_body_bindings` records every `:=`/`var` binding with its start byte, each wiring call filters to bindings preceding the call and takes the LAST one, and the shadow guard likewise only considers bindings in scope at the call position (receiver/parameters always in scope). RED tests `test_earlier_nested_decoy_binding_does_not_win` and `test_shadow_after_wiring_does_not_suppress` demonstrate both the false positive and the false negative the flat snapshot caused.

File: codebase_rag/parsers/rpc_exposure.py [hunk 3]
```diff
@@ -0,0 +1,315 @@
...
+                            elem.child_by_field_name(cs.FIELD_NAME)
+                        )
+                    )
+                ]
+        return []
+
+    def _emit_exposure(self, impl_qn: str, stem: str, method: str) -> None:
+        method_qn = f"{impl_qn}{cs.SEPARATOR_DOT}{method}"
+        if self._function_registry.get(method_qn) is not NodeType.METHOD:
+            return
```
[the remark this answers is not in the document]
Comment (vitali87) [reply codebase_rag/parsers/rpc_exposure.py]: Pushing back on this one. The wiring call `New<Stem>Handler(impl)` only compiles when `impl` satisfies the generated interface, so the Go compiler has already performed exactly the signature check this suggests; re-deriving it graph-side would duplicate the compiler using bare type-name strings that cannot faithfully model cross-package parameter types, and any mismatch in that approximation would silently drop genuine exposures (a false negative in the direction this feature exists to fix). The impl type itself is not guessed by name: it is resolved from the actual wiring argument (literal type, constructor return type, or typed parameter), so a same-named-but-wrong-signature method on an UNRELATED type never enters the lookup. Resolving as not planned.

**Comment by vitali87** [comment]
@greptile review

File: codebase_rag/parsers/rpc_exposure.py [hunk 4]
```diff
@@ -0,0 +1,526 @@
...
+                            for c in child.named_children
+                            if c.type == cs.TS_GO_PACKAGE_IDENTIFIER
+                        ),
+                        None,
+                    )
+                    if name := go_utils.type_identifier_text(child):
+                        entries.append((qualifier, name))
+                    break
+                if name := go_utils.type_identifier_text(child):
+                    entries.append((None, name))
```
[the remark this answers is not in the document]
Comment (vitali87) [reply codebase_rag/parsers/rpc_exposure.py]: This one does not reproduce on our grammar. In the vendored tree-sitter-go, an embedded `*pkg.Base` parses as `field_declaration > (anonymous *) > qualified_type`, so `qualified_type` IS a direct named child and the qualifier branch fires; there is no pointer_type wrapper to bypass it. Verified with a direct AST dump and locked in with the pipeline regression `test_pointer_embedded_cross_package_base_exposes` (cross-package `*base.Base` promotion), which passed BEFORE any change in e042ae2f and stays as coverage. Resolving as not-reproducible, with the regression test guarding the case.

File: codebase_rag/parsers/rpc_exposure.py [hunk 5]
```diff
@@ -0,0 +1,526 @@
...
+                module_qn, arg, position, param_types, bindings
+            )
+            if impl_qn is None:
+                continue
+            connect_dir = import_map.get(qualifier)
+            if connect_dir is None:
+                continue
+            for method in self._contract_methods(connect_dir, stem):
+                if source_qn := self._method_source_qn(impl_qn, method, set()):
+                    self._emit_exposure(source_qn, stem, method)
```
[the remark this answers is not in the document]
Comment (vitali87) [reply codebase_rag/parsers/rpc_exposure.py]: Fixed in e042ae2f. Bindings now carry the byte range they are visible in (`_Binding.scope_end` = the enclosing block's end), and both the shadow guard and impl resolution use `_binding_at`, which requires `start < position < scope_end`. A shadow inside an earlier, already-closed nested block no longer suppresses a genuine wiring call. RED test `test_shadow_in_closed_block_does_not_suppress` reproduced the false negative before the fix.

**Comment by vitali87** [comment]
@greptile review

File: codebase_rag/parsers/rpc_exposure.py [hunk 6]
```diff
@@ -0,0 +1,591 @@
...
+                    != type_name
+                ):
+                    continue
+                return GoRpcExposureProcessor._struct_embedded_entries(struct)
+        return []
+
+    @staticmethod
+    def _struct_embedded_entries(struct: Node) -> list[tuple[str | None, str]]:
+        entries: list[tuple[str | None, str]] = []
+        field_list = next(
```
[the remark this answers is not in the document]
Comment (vitali87) [reply codebase_rag/parsers/rpc_exposure.py:547]: Fixed in 7b742948. The traversal now separates the two concerns: `visited` guards cycles only (unbounded, since embeds are finite and cycle-checked), and a dedicated `depth` parameter caps chain length at 4 without sibling breadth counting against it. RED test `test_many_embedded_siblings_do_not_exhaust_promotion` (nine filler embeds before the defining one) reproduced the dropped edge before the fix and is green after.

**Comment by vitali87** [comment]
@greptile review

File: codebase_rag/parsers/rpc_exposure.py [hunk 7]
```diff
@@ -0,0 +1,594 @@
...
+    ) -> str | None:
+        # The node serving a contract method: defined on the impl type
+        # directly, or promoted from an embedded type (Go embedding is not
+        # inheritance, so the graph has no parent edge to follow). Promoted
+        # stubs from a generated `*connect` package (`Unimplemented<Stem>
+        # Handler`) are not served RPCs. `visited` guards cycles; the DEPTH
+        # cap bounds chains without counting sibling breadth against it.
+        if impl_qn in visited or depth > 4:
+            return None
+        visited.add(impl_qn)
```
[the remark this answers is not in the document]
Comment (vitali87) [reply codebase_rag/parsers/rpc_exposure.py]: Fixed in 7bec0caa. `visited` is now a map of type to the shallowest depth it was traversed at, and re-traversal is allowed at a strictly shallower depth, so a depth-exhausted diamond branch no longer suppresses a valid shallow path while cycles and repeats still terminate. RED test `test_diamond_embedding_finds_shallow_path` (Impl embedding both a five-deep chain and its tail directly) reproduced the lost edge before the fix.

**Comment by vitali87** [comment]
@greptile review

File: codebase_rag/parsers/rpc_exposure.py [hunk 8]
```diff
@@ -0,0 +1,597 @@
...
+    @staticmethod
+    def _embedded_entry(decl: Node) -> tuple[str | None, str] | None:
+        for child in decl.named_children:
+            qualifier = None
+            if child.type == cs.TS_GO_QUALIFIED_TYPE:
+                qualifier = next(
+                    (
+                        safe_decode_text(c)
+                        for c in child.named_children
+                        if c.type == cs.TS_GO_PACKAGE_IDENTIFIER
```
[the remark this answers is not in the document]
Comment (vitali87) [reply codebase_rag/parsers/rpc_exposure.py:573]: Fixed in 735a9722 at the root: the arbitrary cap is gone entirely. Go places no limit on promotion depth, and the min-depth visited map already guarantees termination on its own (a type re-traverses only at a strictly shallower depth, so recorded depths only ever decrease and path cycles are blocked). RED test `test_deep_promotion_chain_exposes` (six-level chain, no shallower path) reproduced the dropped edge before the fix; the diamond and breadth regressions stay green.

**Comment by vitali87** [comment]
@greptile review
